### PR TITLE
PLT-359 ADR regarding problem with PAB's integration with indexing solutions.

### PR DESCRIPTION
**Pull request by koslambrou**
ADR regarding problem with PAB's integration with indexing solutions.

This was triggered by the issue #473. A temporary fix was submitted by #544, but the ideal solution should be the one presented in this ADR.

The ADR is found here: https://plutus-apps--550.org.readthedocs.build/en/550/adr/0004-pab-indexing-solution-integration.html

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [ ] Reviewer requested

### CODE REVIEW

File: doc/adr/0002-pab-indexing-solution-integration.rst [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+Pros:
+
+- No modification to the Contract API
+- All the indexed information is consistent with each other
+
+Cons:
+
+- Hard to implement
+- As fast as the slowest indexing solution
+- Can not be feasible with external indexing solutions which we can’t control the indexing speed (e.g. Blockfrost_ or Scrolls_)
```
Comment (joseph-fajen) [inline doc/adr/0002-pab-indexing-solution-integration.rst:99]: suggest adding "for" --> indexing solutions for which we can't control

File: doc/adr/0002-pab-indexing-solution-integration.rst [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+Pros:
+
+- Limits design change on the PAB
+- More control given to the user of the Contract API
+
+Cons:
+
+- Adds (undesirable?) complexity to the Contract API
+- We'll need to add a bunch of functions (ex. `awaitMarconiTxConfirmed or `awaitScrollsTxConfirmed`) for each new indexing solution we want to support
```
Comment (joseph-fajen) [inline doc/adr/0002-pab-indexing-solution-integration.rst:114]: recommend using either "e.g." or "for example" instead of ex. --> 

(e.g., 'awaitMarconiTxConfirmed' or ... ) 
(for example, 'awaitMarconiTxConfirmed' or ... )

**Review by joseph-fajen — APPROVED** [review]
I had just a few minor comments. Looks good!

File: doc/adr/0002-pab-indexing-solution-integration.rst [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+Pros:
+
+- No modification to the Contract API
+- All the indexed information is consistent with each other
+
+Cons:
+
+- Hard to implement
+- As fast as the slowest indexing solution
+- Can not be feasible with external indexing solutions for which we can’t control the indexing speed (e.g. Blockfrost_ or Scrolls_)
```
Comment (raduom) [inline doc/adr/0002-pab-indexing-solution-integration.rst:99]: While true, I don't think that anyone would expect that we would synchronise with external indexing solutions.

File: doc/adr/0002-pab-indexing-solution-integration.rst [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+Cons:
+
+- Adds (undesirable?) complexity to the Contract API
+- We'll need to add a bunch of functions (e.g., `awaitMarconiTxConfirmed or `awaitScrollsTxConfirmed`) for each new indexing solution we want to support
+
+Query functions should interact with a single source of truth
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this scenario, we make the design decision that the Contract API should only interact with a single indexing solution. Thus, any blockchain information currently stored in the PAB should be moved to the indexing solution. Also, combining indexing solutions would need to be integrated in the single indexing solution that’s connected to the PAB.
```
Comment (raduom) [inline doc/adr/0002-pab-indexing-solution-integration.rst:119]: This makes the most sense.

File: doc/adr/0002-pab-indexing-solution-integration.rst [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+Pros:
+
+- Simplest in design to implement (other than manual work to move code)
+- No modification to the Contract API
+
+Cons:
+
+- PAB won't be able to work with external indexing solutions (e.g. Blockfrost_ or Scrolls_)
+
+Implications
```
Comment (raduom) [inline doc/adr/0002-pab-indexing-solution-integration.rst:130]: There is one other option I think, namely: Change the query APIs to include some sort of ValidSlots information. If the currently indexed slot of the ChainIndex/PAB is not up to what we need either wait for it or error out.

The advantage of this solution is that it would work across different indexers as long as we can get an indication as to the slot number they are up to (which is not unreasonable to assume), it would be backward compatible (make new queries).

Disadvantages are that the implementation would be a bit complicated and it's very similar to what I think we'll do with Marconi (it's the simplest way of achieving some sort of synchronisation across indexers).

**Review by raduom — APPROVED** [review]
LGTM

File: doc/adr/0002-pab-indexing-solution-integration.rst [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+In this scenario, we make the design decision that the Contract API should only interact with a single indexing solution. Thus, any blockchain information currently stored in the PAB should be moved to the indexing solution. Also, combining indexing solutions would need to be integrated in the single indexing solution that’s connected to the PAB.
+
+Pros:
+
+- Simplest in design to implement (other than manual work to move code)
+- No modification to the Contract API
+
+Cons:
+
+- PAB won't be able to work with external indexing solutions (e.g. Blockfrost_ or Scrolls_)
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0002-pab-indexing-solution-integration.rst:128]: Right! I'll rephrase it to: "PAB won't be able to integrate with external indexing solutions (e.g. Blockfrost_ or Scrolls_), but these external solutions would need to be integrated into the single source of truth indexing solution. Therefore, this solution reduces the PAB's flexibility.`

File: doc/adr/0002-pab-indexing-solution-integration.rst [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+Pros:
+
+- Limits design change on the PAB
+- More control given to the user of the Contract API
+
+Cons:
+
+- Adds (undesirable?) complexity to the Contract API
+- We'll need to add a bunch of functions (e.g., `awaitMarconiTxConfirmed or `awaitScrollsTxConfirmed`) for each new indexing solution we want to support
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0002-pab-indexing-solution-integration.rst:121]: Added a phrase saying that this option is not an option.

**Review by ak3n — COMMENTED** [review]
The overall impression is that we have a choice between two approaches:

1. To hide the synchronization details (slot information of different components) in the indexing part or behind the methods with smart waiting.
2. To provide the details and let the user decide how to deal with the synchronization of states.

I suppose the best way to have both of them, as it depends on the use cases. At the beginning, while learning or for simple use cases, it might be enough the standard solution, but later more flexibility might be needed (but which will be too complex at the beginning).

File: doc/adr/0002-pab-indexing-solution-integration.rst [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+Also, combining indexing solutions would need to be integrated in the single indexing solution that’s connected to the PAB.
+
+Pros:
+
+- Simplest in design to implement (other than manual work to move code)
+- No modification to the Contract API
+
+Cons:
+
+- PAB won't be able to integrate with external indexing solutions (e.g. Blockfrost_ or Scrolls_), but these external solutions would need to be integrated into the single source of truth indexing solution. Therefore, this solution reduces the PAB's flexibility.
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0002-pab-indexing-solution-integration.rst:139]: > [...] (should we show our preference here in comments?)

Good points about preference. I think it's okay to leave them here as comments. I'll add the PR in the "Notes" section of the ADR.

> It separates the semantics of the components. Indexing happens in one place, running contracts in another one.

Does this contradict what you said in the previous comment? You say here that you prefer the "single source of truth" approach, but in your previous comment you say that we should allow (eventually) users to deal with state synchronization of components. But if we have a single source of truth, then we don't need to deal with state synchronization of components.

[the remark this answers is not in the document]
Comment (ak3n) [reply doc/adr/0002-pab-indexing-solution-integration.rst:139]: Yeah, at first I wrote the comment in this thread, choosing between the proposed solutions.

And later I tried to reflect about some general solution, what approach is better for us and how it fits the PAB's vision I suppose.

[the remark this answers is not in the document]
Comment (andreabedini) [reply doc/adr/0002-pab-indexing-solution-integration.rst:139]: IMHO whathever indexing solution PAB uses should be hidden from the user. So `unspentTxOutFromRef` (~ a PAB function) should be in charge of making sure plutus-chain-index has caught up with the rest of PAB.

[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0002-pab-indexing-solution-integration.rst]: @andreabedini I'm interpreting this as solution 1 of this ADR (Make sure all components are in sync), but it might be different. You're saying that all calls to PAB functions in the Contract Monad should wait for all other components (like chain-index) to be synced and, similarly, calls to chain-index funcions should wait for all other components to be synced.

[the remark this answers is not in the document]
Comment (andreabedini) [reply doc/adr/0002-pab-indexing-solution-integration.rst]: @koslambrou isn't it "Query functions should interact with a single source of truth" ? maybe I am missing some details.
> You're saying that all calls to PAB functions in the Contract Monad should wait for all other components (like chain-index) to be synced and, similarly, calls to chain-index funcions should wait for all other components to be synced.

Yes, this is what I am thinking.

[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0002-pab-indexing-solution-integration.rst]: If that's what you're thinking, then you're talking about the category "Make sure all components are in sync". With the "Query functions should interact with a single source of truth" you don't need to worry about functions not being in sync, because there is only ONE component which indexes information (all the indexing information from the PAB would be moved to that single source of truth indexing component).

File: doc/adr/0004-pab-indexing-solution-integration.rst [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+If the currently indexed slot of the PAB and indexing component is not up to what we need, we can either wait for it or throw an error.
+
+Pros:
+
+- Would work accross different indexers as long as we can get an indication as to the slot number they are synced up to (which is not unreasonable to assume)
+- Limits the changes to the Contract API
+
+Cons:
+
+- Makes the query APIs more complicated, which ultimately result in a more complex contract
```
Comment (raduom) [inline doc/adr/0004-pab-indexing-solution-integration.rst:171]: I am not sure this is the case. The contract API consists of a EDSL (effect system based), so we could implement synchronisation as a cross-cutting/semantic/interpreter concern, by specifying a 'context configuration' where the user would decide what is acceptable for their application in general.

File: doc/adr/0002-pab-indexing-solution-integration.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreabedini) [reply doc/adr/0002-pab-indexing-solution-integration.rst]: @koslambrou The way I see it is: from the POV of the contract monad, there is only one source of truth, the interpreter which is run by PAB. The interpreter giving inconsistent results is, IMHO, 100% an interpreter bug.

File: doc/adr/0004-pab-indexing-solution-integration.rst [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+* We will replace `plutus-chain-index` with `Marconi` as PAB's indexing component
+
+* We will move out the Blockchain information indexed by PAB in Marconi
+
+* We will add new indexers in Marconi in order to replicate the information indexed by `plutus-chain-index`
+
+* We will adapt the architecture of Marconi (which will become our new indexing component) to support waiting queries
+
+* Since HTTP APIs are not well suited to waiting queries, we will use RPC via UNIX Socket to communicate information between PAB and Marconi
```
Comment (andreabedini) [inline doc/adr/0004-pab-indexing-solution-integration.rst:93]: Why not using Marconi has a library?

File: doc/adr/0004-pab-indexing-solution-integration.rst [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+We could, before querying, wait for `plutus-chain-index` to be in sync with all components which index blockchain data (PAB, Blockfrost_).
+
+Pros:
+
+- All the indexed information is consistent with each other
+
+Cons:
+
+- As fast as the slowest indexing component
+- Tight coupling between the indexing components meaning that if the Contract only uses chain-index requests without using requests from other indexing components, the chain-index will still have to wait for all other components to be in sync with each other
```
Comment (andreabedini) [inline doc/adr/0004-pab-indexing-solution-integration.rst:119]: maybe the interpreter can be given the list of relevant indexers?

File: doc/adr/0004-pab-indexing-solution-integration.rst [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+Pros:
+
+- Simplest in design to implement (other than manual work to move code)
+- No modification to the Contract API
+- Augments PAB's cohesion, because it's responsability will be limited to contract instance management
+
+Cons:
+
+- PAB won't be able to integrate with external indexing components (e.g. Blockfrost_ or Scrolls_), but these external solutions would need to be integrated into the single source of truth indexing solution. Therefore, this solution reduces the PAB's flexibility.
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply doc/adr/0004-pab-indexing-solution-integration.rst:153]: > PAB won't be able to integrate with external indexing components
Was this ever in our roadmap? If not, it should not be listed either.

File: doc/adr/0002-pab-indexing-solution-integration.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreabedini) [reply doc/adr/0002-pab-indexing-solution-integration.rst]: Btw, I agree with the decisions in this ADR so perhaps we are just using different words to express the same concept.

File: doc/adr/0004-pab-indexing-solution-integration.rst [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+* We will adapt the architecture of Marconi (which will become our new indexing component) to support waiting queries
+
+* Since HTTP APIs are not well suited to waiting queries, we will use RPC via UNIX Socket to communicate information between PAB and Marconi
+
+Alternative solutions
+---------------------
+
+In this section, we describe all the ways to deal with the problem.
+Note that final decision might include one or more of these alternate solutions.
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0004-pab-indexing-solution-integration.rst:99]: :D Totally write. I'll reword this or simply remove it.

File: doc/adr/0004-pab-indexing-solution-integration.rst [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+The best solution is probably a combination of the `Alternative solutions`_ described below.
+However, we will mainly choose the `Query functions should interact with a single source of truth`_ solution.
+
+* We will replace `plutus-chain-index` with `Marconi` as PAB's indexing component
+
+* We will move out the Blockchain information indexed by PAB in Marconi
+
+* We will add new indexers in Marconi in order to replicate the information indexed by `plutus-chain-index`
+
+* We will adapt the architecture of Marconi (which will become our new indexing component) to support waiting queries
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0004-pab-indexing-solution-integration.rst:91]: Currently, we have only implemented indexers which stores information in the database. But the indexer also have a notification field, right? Then this notification field allows an indexer to support waiting queries (like `awaitTxConfirmed`).

I used the word `adapt` because we haven't officially made a decision on how the notification system will work. How would you rephrase it?

File: doc/adr/0002-pab-indexing-solution-integration.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0002-pab-indexing-solution-integration.rst]: > @koslambrou The way I see it is: from the POV of the contract monad, there is only one source of truth, the interpreter which is run by PAB. The interpreter giving inconsistent results is, IMHO, 100% an interpreter bug.

Good point! I understand what you meant now.

File: doc/adr/0004-pab-indexing-solution-integration.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0004-pab-indexing-solution-integration.rst:153]: It's not in the official roadmap yet, but there is demand for using other indexing components with the PAB (`Blockfrost` is one of them).

[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0004-pab-indexing-solution-integration.rst:153]: > I am not sure this is a con. Reducing a component's responsibilities seems like a pro to me.

Makes sense. High cohesion is always good :D

File: doc/adr/0004-pab-indexing-solution-integration.rst [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+Pros:
+
+- Simplest in design to implement (other than manual work to move code)
+- No modification to the Contract API
+- Augments PAB's cohesion, because it's responsability will be limited to contract instance management
+
+Cons:
+
+- PAB won't be able to integrate with external indexing components (e.g. Blockfrost_ or Scrolls_), but these external solutions would need to be integrated into the single source of truth indexing solution. Therefore, this solution reduces the PAB's flexibility.
+- The design of the indexing component will need to be changed to support waiting queries (like the `awaitTxConfirmed` from PAB)
```
[the remark this answers is not in the document]
Comment (koslambrou) [reply doc/adr/0004-pab-indexing-solution-integration.rst:154]: In this case, I'm talking about "indexing component" in general, not specifically Marconi. I'm basically saying that the indexing components needs to support waiting queries (which Marconi will). But if someone wants to replace PAB's main indexing component with another one, it needs to support waiting queries.

File: doc/adr/0004-pab-indexing-solution-integration.rst (hunk 10 above)
Comment (koslambrou) [reply doc/adr/0004-pab-indexing-solution-integration.rst:93]: Mmmm... very good point. I guess if we're using RPC via UNIX Socket, we suppose the Marconi executable is on the same machine as the PAB, therefore it makes totally to use it as a library.

File: doc/adr/0004-pab-indexing-solution-integration.rst (hunk 11 above)
Comment (koslambrou) [reply doc/adr/0004-pab-indexing-solution-integration.rst:119]: Problem with that is that the PAB doesn't know what the "relevant" indexers are (unless they are specified by the user when starting the PAB). However, doing it that way will make the PAB highly coupled with the builtin contracts. Might not be such a problem right now, but imagine we extend the PAB to be even more generalized, where the contracts are not builtin, but are actually separate executable files which can be injected in PAB.

Therefore, I believe PAB should not care about what requests the contract is composed of.

File: doc/adr/0004-pab-indexing-solution-integration.rst (hunk 9 above)
Comment (koslambrou) [reply doc/adr/0004-pab-indexing-solution-integration.rst:171]: True! But the downside of this is that is doesn't actually solve the problem in the context right? The contract would throw a runtime error if the chain-index is not fully synced. Compare that to the "single source of truth" solution were you would never get a synchronization error.

The upside is that the error would be very clear and explicit.

File: doc/adr/0004-pab-indexing-solution-integration.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (raduom) [reply doc/adr/0004-pab-indexing-solution-integration.rst:91]: The phrasing is fine, I was just curious what you meant. You are making it sound a bit like a teaser of things to come in the next episode. Which is fine by me. :)

File: doc/adr/0004-pab-indexing-solution-integration.rst (hunk 9 above)
Comment (raduom) [reply doc/adr/0004-pab-indexing-solution-integration.rst:171]: As @andreabedini noticed before, it's all about the implementation of the interpreter. You can have an interpreter that waits until all indexers have caught-up with the chain in which case you would never get a synchronisation error either.
